src/twisted/internet/defer.py

```diff
@@ -1393,24 +1535,30 @@ class _CancellationStatus:
     """
     Cancellation status of an L{inlineCallbacks} invocation.
 
-    @ivar waitingOn: the L{Deferred} being waited upon (which
-        L{_inlineCallbacks} must fill out before returning)
-
     @ivar deferred: the L{Deferred} to callback or errback when the generator
         invocation has finished.
+    @ivar waitingOn: the L{Deferred} being waited upon (which
+        L{_inlineCallbacks} must fill out before returning)
     """
 
-    deferred = attr.ib()
-    waitingOn = attr.ib(default=None)
+    deferred = attr.ib()  # type: Deferred[object]
+    waitingOn = attr.ib(default=None)  # type: Optional[Deferred[object]]
 
 
-@failure._extraneous
-def _inlineCallbacks(result, g, status):
+@_extraneous
+def _inlineCallbacks(
+    result: object,
+    gen: Union[
+        Generator[Deferred[_T], object, None],
+        Coroutine[Deferred[_T], object, None],
+    ],
```

Comment:
Repeating a comment from @mstojcevich in @glyph's suggestion PR here so we don't lose it:

> Is None correct for the generator/coroutine return type for these? I think it should be Optional[object] since this handles pulling the value off of StopIteration(x) when the generator returns x.
> 
> And I don't think YieldType is necessarily a Deferred when gen is a Generator.
> 
> I think the same concerns apply to the original PR fwiw.